Bluetooth: host: Fix host send cmd complete to dis-conn

LingaoM opened the pull request:
Fixes: #38054

Signed-off-by: Lingao Meng <menglingao@xiaomi.com>

LingaoM committed:
Bluetooth: host: Fix host send cmd complete to dis-conn

Fixes: #38054

Signed-off-by: Lingao Meng <menglingao@xiaomi.com>

Thalley reviewed (changes requested):
See my comment in the issue: https://github.com/zephyrproject-rtos/zephyr/issues/38054#issuecomment-908354699

LingaoM committed:
Bluetooth: host: Fix hci command done

When received hci command rsp which not for expected
hci command, we should ignore this.

Signed-off-by: Lingao Meng <menglingao@xiaomi.com>

LingaoM commented:
@joerchan @Thalley Please take a look.

Thalley commented:
> @joerchan @Thalley Please take a look.

@LingaoM The discussion in the issue that is supposedly fixed by this PR are pointing towards a controller issue, so we need to resolve that before we can close the issue based on this PR.

LingaoM commented:
But I think my modification still makes sense.

Thalley commented on subsys/bluetooth/host/hci_core.c:
```diff
@@ -193,7 +193,8 @@ void bt_hci_host_num_completed_packets(struct net_buf *buf)
 	}
 
-	if (!bt_conn_is_handle_valid(conn)) {
+	if (conn->state != BT_CONN_CONNECTED &&
+	    conn->state != BT_CONN_DISCONNECT) {
```
This change here is effectively removing `BT_CONN_DISCONNECT_COMPLETE` from entering the clause, right? 

Can you elaborate on why the BT_CONN_DISCONNECT_COMPLETE should not be able to enter the clause?

carlescufi replied:
@LingaoM can you please reply?

carlescufi replied:
This is fixing a) in https://github.com/zephyrproject-rtos/zephyr/issues/38054#issuecomment-911518510

Thalley commented on subsys/bluetooth/host/hci_core.c:
```diff
@@ -2008,8 +2008,11 @@ static void hci_cmd_done(uint16_t opcode, uint8_t status, struct net_buf *buf)
+		return;
 	}
 
+	bt_dev.sent_cmd = NULL;
+
```
Can you elaborate on why this statement was moved? It isn't clear to me.

LingaoM replied:
see above.

Thalley reviewed (changes requested):
The first commit, nor the PR, fixes the reported issue. 

Please also update the commit message of the first commit, and remove the `fixes` keyword.

carlescufi commented on subsys/bluetooth/host/hci_core.c:
```diff
@@ -2007,8 +2008,11 @@ static void hci_cmd_done(uint16_t opcode, uint8_t status, struct net_buf *buf)
 	if (cmd(buf)->opcode != opcode) {
 		BT_WARN("OpCode 0x%04x completed instead of expected 0x%04x",
 			opcode, cmd(buf)->opcode);
+		return;
```
This is fixing b) in https://github.com/zephyrproject-rtos/zephyr/issues/38054#issuecomment-911518510

carlescufi commented on subsys/bluetooth/host/buf.c:
```diff
@@ -85,7 +85,6 @@ struct net_buf *bt_buf_get_cmd_complete(k_timeout_t timeout)
 
 	key = irq_lock();
 	buf = bt_dev.sent_cmd;
-	bt_dev.sent_cmd = NULL;
```
Why is this needed?

LingaoM replied:
the cmd will be clear only when corresponding response received, so when we received unexpected response(host set number completed packet response) we can also continue to receive response.

Thalley replied:
@LingaoM It looks like `send_cmd` may also clear `bt_dev.sent_cmd` and set it to a new command. Is there any way that we will clear the wrong `bt_dev.sent_cmd` here?

LingaoM replied:
- host send host number of complete package command.

- host send scan param set command

- Controller response command complete for host number of complete command.

- Controller response command complete for scan param set command.

LingaoM replied:
send_cmd will be clear by received `Controller response command complete for host number of complete command.`, but this not we expect to do.

Thalley replied:
> Controller response command complete for host number of complete command.

The controller won't normally send this, unless there is an error, right? 

I am not sure what the potential side effects of moving this is, or what it fixes, so I vary about approving it.

LingaoM replied:
@Thalley @rugeGerritsen

Thalley replied:
Sorry for the late response @LingaoM  but I've been trying to really follow your comment and how this variable is used, when it is cleared and when it is set, and I still haven't been able to wrap my head around it (even after looking at this over several days). Since I still don't completely follow the change, I can't approve of this just yet, but I won't block it if e.g. @rugeGerritsen and @joerchan  can sign off on it.

LingaoM replied:
@Thalley I need to explain the reasons in detail:

We have two types of Host send command. One is to send immediately, the other is to send sequentially.

For _BT_HCI_OP_HOST_NUM_COMPLETED_PACKETS_ directly call **bt_send** to send (refer to **bt_hci_cmd_send** https://github.com/zephyrproject-rtos/zephyr/blob/main/subsys/bluetooth/host/hci_core.c#L265).

For other opcodes, they are put into **cmd_tx_queue**, and then processed in the task of **hci_tx_thread::send_cmd**.

The **sent cmd** will be recorded through **bt_dev.send_cmd** (https://github.com/zephyrproject-rtos/zephyr/blob/main/subsys/bluetooth/host/hci_core.c#L2342).
Which is used to locate the corresponding command when command complete or command status is received in (https://github.com/zephyrproject-rtos/zephyr/blob/main/subsys/bluetooth/host/buf.c#L97).

At that time, we made a mistake here that we took for granted.
We believe that the cmd response we received at this time was generated by the command we just sent(https://github.com/zephyrproject-rtos/zephyr/blob/main/subsys/bluetooth/host/buf.c#L97). 

As I said in the previous commit, this is not exactly the case:

As shown in the screenshot of issure #38054 :

After Host send _host number complete packets_, _LE Set Scan Params_ was sent **immediately** and at this time, our `bt.dev.send_cmd=buf(LE Set Scan Params)`.

Then we received _HCI Command Complete ( Invalid HCI Command params for BT_HCI_OP_HOST_NUM_COMPLETED_PACKETS)_, 
this is not what we expect to receive, so it should be discarded(https://github.com/zephyrproject-rtos/zephyr/pull/38056/files#diff-774b3de86adb275da54becf2a0958a2d4dae03d63e47840037ecc575e9d658b4R2011).

I think we should continue to be able to receive the Command complete corresponding to _LE Set Scan Params_, but if we are receiving HCI After Command Complete (Invalid HCI Command params for BT_HCI_OP_HOST_NUM_COMPLETED_PACKETS), set **bt_dev.send_cmd=NULL**(https://github.com/zephyrproject-rtos/zephyr/blob/main/subsys/bluetooth/host/buf.c#L98), 

Then the subsequent HCI Command complete will not be processed correctly (because this response cannot find the corresponding Command first(https://github.com/zephyrproject-rtos/zephyr/blob/main/subsys/bluetooth/host/hci_core.c#L2003).

Which eventually leads to **k_sem_take failed with err** in https://github.com/zephyrproject-rtos/zephyr/blob/main/subsys/bluetooth/host/hci_core.c#L304).

LingaoM replied:
@carlescufi @joerchan Please also take a look.

Thalley replied:
Thanks @LingaoM I think I get it now :)

rugeGerritsen commented:
@LingaoM , would you mind explaining how you have tested this change?

rugeGerritsen commented:
> @LingaoM , would you mind explaining how you have tested this change?

@LingaoM , can you comment on this?

LingaoM commented:
> > @LingaoM , would you mind explaining how you have tested this change?
> 
> @LingaoM , can you comment on this?

see https://github.com/zephyrproject-rtos/zephyr/pull/38056#discussion_r701781444

rugeGerritsen commented on subsys/bluetooth/host/hci_core.c:
```diff
@@ -193,7 +193,8 @@ void bt_hci_host_num_completed_packets(struct net_buf *buf)
 
-	if (!bt_conn_is_handle_valid(conn)) {
+	if (conn->state != BT_CONN_CONNECTED &&
+	    conn->state != BT_CONN_DISCONNECT) {
```
Some controllers may wait with raising the DISCONNECTED_EVT until all received data buffers are forwarded to the host. 
Not sending HCI Host Num Completed command may result in a deadlock. 

Therefore I believe we should not validate against BT_CONN_DISCONNECT here

LingaoM replied:
You mean only this ?

``` C
if (conn->state != BT_CONN_CONNECTED
```

LingaoM replied:
@Thalley Please take a look.

rugeGerritsen replied:
> You mean only this ?
> 
> ```c
> if (conn->state != BT_CONN_CONNECTED
> ```

@LingaoM, yes. It does not make a difference if the host tries to be clever. It anyways need to handle the case where the controller returns an error code

LingaoM replied:
> Some controllers may wait with raising the DISCONNECTED_EVT until all received data buffers are forwarded to the host.
Not sending HCI Host Num Completed command may result in a deadlock.

But my understanding is that if we do not send host complete in this state(BT_CONN_DISCONNECT), will it cause a deadlock?

So there is nothing wrong with my modification.

Thalley replied:
I am not 100% sure I follow @rugeGerritsen comments. Ruge, can you elaborate a bit?

rugeGerritsen replied:
This is a scenario that can happen:

1. Host calls _HCI_Host_Buffer_Size_ and informs the controller it has one available buffer.
2. Controller receives data from the peer device and sends it to the host. From now on the controller cannot forward data to the host until it has called _HCI_Host_Number_Of_Completed_Packets_.
3. The controller receives another data packet from the peer device. It waits with forwarding it to the host.
4. The host calls _HCI_Disconnect_. The controller waits with forwarding the _HCI_Disconnection_Complete_ event as it still has more data to the host.
5. The host calls _HCI_Host_Number_Of_Completed_Packets_. The controller now forwards the pending data. Then it raises the _HCI_Disconnection_Complete_ event.

Therefore, if we make the host stop sending _HCI_Host_Number_Of_Completed_Packets_ after the  _HCI_Disconnect_ command, the host will never receive the _HCI_Disconnection_Complete_ event.

LingaoM replied:
Under my current modification, even in this state(`BT_CONN_DISCONNECT`), which cause by host call `bt_conn_disconnect`, `HCI_Host_Number_Of_Completed_Packets` can be sent.

Thalley replied:
Agree with @LingaoM : In the change here, the host will continue to send completed packets in the BT_CONN_CONNECTED and BT_CONN_DISCONNECT states.

rugeGerritsen replied:
Thanks, you are right.

Thalley reviewed (approved):
LGTM, but I think we should get at least one BT host code owner to approve such an integral change as well.

carlescufi commented:
@jhedberg a review from you on this one would be appreciated.

jhedberg reviewed (changes requested):
There seems to be some missing buffer reference counting changes. Previously the code would “take” the reference owned by bt_dev.sent_cmd which is why it was important to immediately set it to NULL. Now that this does not happen you should be using net_buf_ref() and be sure to unref when finally setting to NULL.

LingaoM commented:
> There seems to be some missing buffer reference counting changes. Previously the code would “take” the reference owned by bt_dev.sent_cmd which is why it was important to immediately set it to NULL. Now that this does not happen you should be using net_buf_ref() and be sure to unref when finally setting to NULL. 

Maybe we unref buffer in there：

``` C

void hci_event_prio(struct net_buf *buf)
{
	struct net_buf_simple_state state;
	struct bt_hci_evt_hdr *hdr;
	uint8_t evt_flags;

	net_buf_simple_save(&buf->b, &state);

	BT_ASSERT(buf->len >= sizeof(*hdr));

	hdr = net_buf_pull_mem(buf, sizeof(*hdr));
	evt_flags = bt_hci_evt_get_flags(hdr->evt);
	BT_ASSERT(evt_flags & BT_HCI_EVT_FLAG_RECV_PRIO);

	handle_event(hdr->evt, buf, prio_events, ARRAY_SIZE(prio_events));

	if (evt_flags & BT_HCI_EVT_FLAG_RECV) {
		net_buf_simple_restore(&buf->b, &state);
	} else {
		net_buf_unref(buf);
	}
}

```

LingaoM commented on subsys/bluetooth/host/buf.c:
```diff
@@ -85,7 +85,6 @@ struct net_buf *bt_buf_get_cmd_complete(k_timeout_t timeout)
 
 	key = irq_lock();
 	buf = bt_dev.sent_cmd;
```
Also should return net_buf_ref(buf)
If send_cmd not null, right? @jhedberg

carlescufi replied:
This line should be:
`buf = net_buf_ref(bt_dev.sent_cmd)`

LingaoM replied:
`bt_dev.sent_cmd` maybe null ?

LingaoM replied:
Done.

LingaoM commented on subsys/bluetooth/host/hci_core.c:
```diff
@@ -2007,8 +2008,11 @@ static void hci_cmd_done(uint16_t opcode, uint8_t status, struct net_buf *buf)
+		return;
 	}
 
+	bt_dev.sent_cmd = NULL;
```
@jhedberg You are right,

We missing call net_buf_unref(bt_dev.send_cmd) here.
